sdks/python/apache_beam/io/avroio.py

```diff
@@ -627,11 +627,19 @@ def write_record(self, writer, value):
     writer.append(value)
 
 
+class _FastAvroWriter(Writer):
+  """An adapter class which exposes a file handle so that it can be closed
+  by the sink. """
+  def __init__(self, file_handle, schema, codec):
+    super(_FastAvroWriter, self).__init__(file_handle, schema, codec)
```

Comment:
What is the context for this change (do you by chance have a pertaining fastavro commit that changes this behavior)?

Reply:
It turned out this change is unnecessary -- I'll remove it then.

The problem was with pyenv and Python interpreter on my machine. It lacked libpython.so. As a result, I couldn't import cythonized code. Instead, a pure Python version, `_write_py.py`, was being used: https://github.com/fastavro/fastavro/blob/master/fastavro/write.py#L3. That pure Python version has some differences, but most importantly, the `Writer` class doesn't set `self.fo` field: 

https://github.com/fastavro/fastavro/blob/master/fastavro/_write_py.py#L411 (pure Python)
vs.
https://github.com/fastavro/fastavro/blob/master/fastavro/_write.pyx#L559 (cython)

I wrongly assumed it was the fault of updating the package.

Reply:
Actually, it is a change in 0.22.0: https://github.com/fastavro/fastavro/pull/343/commits/6c55c318735278d097b05da7077b23fead50a039#diff-96a4a8649900b3ff9f22a57de25baca4R498
https://github.com/fastavro/fastavro/blob/f468264b0f86661f1893c8b32bc592fb44e39c75/ChangeLog#L91

Cython is an optional dependency in Beam, so we should not have to rely on presence of cython.

It seems strange that fastavro public api is now different in cythonized and non-cythonized version. Are you interested in reaching out to fastavro folks to clarify whether this is something they want to fix?

We explicitly close the filehandle due to: https://issues.apache.org/jira/browse/BEAM-7747. Perhaps there is a better way?

Reply:
Asked on https://github.com/fastavro/fastavro/pull/343#discussion_r497095926

Reply:
Thanks. We've already got an answer, let's wait until we know if it's possible for the `fo` attribute to become available.

Reply:
Let's do the following:

1. File a JIRA for this issue, mention in the the error message that you observed. Let's mention in the JIRA  the workaround (install cython or downgrade fastavro to 0.21.24) in case users discover this error.

2. Add a workaround in code:
```
try:
  writer.fo.close()
except AttributeError:
  # We can clean this up after fastavro lower bound includes  a fix for https://github.com/fastavro/fastavro/issues/472.
  writer.encoder.close()
```
3. Test that the workaround works.

Thanks for spotting this.  How did you catch it by the way? Was it a unit test failure that occurred locally? I wonder why our non-cython test suite does not catch it.

Reply:
> Was it a unit test failure that occurred locally?

Yes.

> I wonder why our non-cython test suite does not catch it.

I think it doesn't really depend on whether cython is installed or not. `pip install fastavro` downloads a wheel built on the specific platform, which already contains libraries that can be imported. For example, on my Linux machine:

```
pip install --force-reinstall fastavro==0.23.6

Collecting fastavro==0.23.6
  Using cached fastavro-0.23.6-cp37-cp37m-manylinux2010_x86_64.whl (1.4 MB)
```

```
ls site-packages/fastavro | grep write 

json_write.py
_logical_writers.cpython-37m-x86_64-linux-gnu.so
logical_writers.py
_logical_writers_py.py
_write.cpython-37m-x86_64-linux-gnu.so   <-- this is the library I'm talking about
write.py
_write_py.py
```

I'm pretty sure the same happens on Jenkins and on our users' machines. So how did I catch the bug? That `.so` library links with `libpython3.x.so`:

```
ldd site-packages/fastavro/_write.cpython-37m-x86_64-linux-gnu.so

linux-vdso.so.1 (0x00007ffd7abf7000)
libpython3.7m.so.1.0 => not found
libc.so.6 => /lib64/libc.so.6 (0x00007f8c548e3000)
/lib64/ld-linux-x86-64.so.2 (0x00007f8c54d07000)
```
which I didn't have on my machine due to misconfiguration (I already know it's pyenv fault, but this is a different story). That's why the first `import` statement failed, `ImportError` was raised, and pure Python version was used instead.

To sum up, it is a problem, but the chances that someone will encounter this are pretty low. Anyway, I think it still makes sense to file a JIRA and test the workaround that you proposed.

Reply:
Ah that's right. Cython presence only matters if  we are  installing fastavro from sources.